rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,456 @@
+) * 100
+```
+
+**Secondary Metrics - Implementation Queries:**
+
+4. **Cache Preservation Effectiveness**
+```promql
+# Worker uptime distribution - more long-lived workers indicate preserved caches
+histogram_quantile(0.90, worker_manager_worker_uptime_seconds_bucket{intended_role="min-capacity"})
+vs
+histogram_quantile(0.90, worker_manager_worker_uptime_seconds_bucket{intended_role="overflow"})
+```
+
+5. **Termination Frequency Impact**
+```promql
+# Rate of minCapacity worker terminations
+rate(worker_manager_min_capacity_worker_terminations_total[1h])
+```
+
+6. **Capacity Response Time**
+```promql
+# Time to fulfill minCapacity requirements after deficit
+avg(worker_manager_min_capacity_deficit_seconds_bucket) by (workerPoolId)
+```
+
+**Success Criteria:**
+- **Task wait time**: 20% reduction in P95 pending→running time
+- **Worker churn**: 50% reduction in worker spawn frequency for minCapacity pools
+- **Cache effectiveness**: 80% of minCapacity workers have >1 hour uptime
+- **Deficit resolution**: <60 seconds average time to resolve minCapacity deficits
+- **Termination cost**: <5% of minCapacity workers terminated for capacity reduction per day
+
+**Trade-off Evaluation:**
+```promql
+# Cost-benefit analysis: cache preservation vs forced terminations
+(
+  # Benefits: Reduced task wait times
+  (baseline_task_wait_time - current_task_wait_time) * tasks_per_hour
+)
+vs
+(
+  # Costs: Cache loss from forced terminations
+  rate(worker_manager_min_capacity_worker_terminations_total{reason="capacity-reduction"}[1h])
+  * avg_cache_warmup_time
+)
+```
+
+## Rollout Strategy
+
+**Phase 1**: Single Pool Testing
+- Enable on one stable test pool (e.g., `gecko-1/decision-gcp` for try branch)
+- Monitor success metrics, cache preservation, and termination frequency
+- Validate that pool benefits outweigh costs of occasional forceful termination
+
+**Phase 2**: Gradual Expansion
+- Roll out to remaining decision pools with stable minCapacity requirements
+- Prioritize pools where minCapacity changes are infrequent
+- Monitor trade-off between cache benefits and capacity management responsiveness
+
+**Phase 3**: Full Deployment
+- Enable for all eligible pools after validating net positive impact
+- Continue monitoring optimization effectiveness across different usage patterns
+- Consider disabling for pools where minCapacity changes are too frequent
+
+## Error Handling and Edge Cases
+
+**Worker Lifecycle Management:**
+- **Pool reconfiguration**: Capacity changes trigger worker replacement, not reconfiguration
+- **Graceful transitions**: When possible, only terminate idle workers to preserve active caches
+- **Resource allocation**: MinCapacity workers mixed with overflow workers on same infrastructure
+
+**Capacity Reduction Strategy:**
+```javascript
+// Hybrid approach: minimize cache loss when reducing capacity
+async handleMinCapacityDecrease(pool, newMinCapacity) {
```

Comment:
ah, here goes this hidden edge-case handling.

I think it should be part of the RFC description, not implementation suggestion

Reply:
Understood, done!